File: 0049-group-anagrams/0049-group-anagrams.js
/**
 * @param {string[]} strs
 * @return {string[][]}
 */
var groupAnagrams = function (strs) {
    const ans = new Map();

    for (const s of strs) {
        const key = s.split("").sort().join("");
        if (!ans.has(key)) {
            ans.set(key, []);
        }
        ans.get(key).push(s);
    }

    return Array.from(ans.values());
};

var strs = ["eat", "tea", "tan", "ate", "nat", "bat"];
console.log(groupAnagrams(strs));

File: 0049-group-anagrams/0049-group-anagrams.py
# time complexity: O(nklogn)
# space complexity: O(n)
from collections import defaultdict
from typing import List


class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        wordMap = defaultdict(list)
        for word in strs:
            key = ''.join(sorted(word))
            wordMap[key].append(word)
        return [row for row in wordMap.values()]

# time complexity: O(nk)
# space complexity: O(nk)
class Solution:
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        result = defaultdict(list)
        for s in strs:
            count = [0] * 26
            for c in s:
                count[ord(c) - ord("a")] += 1
            result[tuple(count)].append(s)
        return list(result.values())

strs = ["eat", "tea", "tan", "ate", "nat", "bat"]
print(Solution().groupAnagrams(strs))
strs = [""]
print(Solution().groupAnagrams(strs))
strs = ["a"]
print(Solution().groupAnagrams(strs))
